File: backend-spring/src/main/java/com/tirthseva/api/service/UserService.java
package com.tirthseva.api.service;

import com.tirthseva.api.dto.user.UserProfileResponse;
import com.tirthseva.api.entity.User;
import com.tirthseva.api.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;

    public List<UserProfileResponse> getAllUsers() {
        return userRepository.findAll().stream()
                .map(this::mapToProfileResponse)
                .collect(Collectors.toList());
    }

    public UserProfileResponse getUserById(Integer id) {
        User user = userRepository.findById(id).orElse(null);
        return user != null ? mapToProfileResponse(user) : null;
    }

    @Transactional
    public boolean deleteUser(Integer id) {
        if (!userRepository.existsById(id)) {
            return false;
        }
        userRepository.deleteById(id);
        return true;
    }

    @Transactional
    public boolean updateUserRole(Integer id, String newRole) {
        // Validate role
        List<String> validRoles = Arrays.asList("User", "ServiceProvider", "Admin");
        if (!validRoles.contains(newRole)) {
            return false;
        }

        User user = userRepository.findById(id).orElse(null);
        if (user == null) {
            return false;
        }

        user.setRole(newRole);
        userRepository.save(user);
        return true;
    }

    private UserProfileResponse mapToProfileResponse(User user) {
        return UserProfileResponse.builder()
                .id(user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .role(user.getRole())
                .isEmailVerified(user.getIsEmailVerified())
                .createdAt(user.getCreatedAt())
                .build();
    }
}

File: backend/TirthSeva.API/Services/UserService.cs
using Microsoft.EntityFrameworkCore;
using TirthSeva.API.Data;
using TirthSeva.API.DTOs;
using TirthSeva.API.Models;

namespace TirthSeva.API.Services
{
    public class UserService
    {
        private readonly ApplicationDbContext _context;

        public UserService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<List<UserProfileResponse>> GetAllUsersAsync()
        {
            return await _context.Users
                .Select(u => new UserProfileResponse
                {
                    Id = u.Id,
                    Name = u.Name,
                    Email = u.Email,
                    Role = u.Role,
                    IsEmailVerified = u.IsEmailVerified,
                    CreatedAt = u.CreatedAt
                })
                .ToListAsync();
        }

        public async Task<UserProfileResponse?> GetUserByIdAsync(int id)
        {
            var user = await _context.Users.FindAsync(id);

            if (user == null)
            {
                return null;
            }

            return new UserProfileResponse
            {
                Id = user.Id,
                Name = user.Name,
                Email = user.Email,
                Role = user.Role,
                IsEmailVerified = user.IsEmailVerified,
                CreatedAt = user.CreatedAt
            };
        }

        public async Task<bool> DeleteUserAsync(int id)
        {
            var user = await _context.Users.FindAsync(id);

            if (user == null)
            {
                return false;
            }

            _context.Users.Remove(user);
            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<bool> UpdateUserRoleAsync(int id, string newRole)
        {
            // Validate role
            if (newRole != "User" && newRole != "ServiceProvider" && newRole != "Admin")
            {
                return false;
            }

            var user = await _context.Users.FindAsync(id);

            if (user == null)
            {
                return false;
            }

            user.Role = newRole;
            await _context.SaveChangesAsync();
            return true;
        }
    }
}
